### PR TITLE
feat: load secrets for OpenStack Helm via a values file

### PR DESCRIPTION
**Pull request by cardoe**
Generate a helm values file with the secrets in it. That can then be stored and used via tools like https://github.com/jkroepke/helm-secrets Then switch to using the OpenStack Helm charts via their repo so that we can execute their helm charts like we do the other components.

### CODE REVIEW

**Comment by cardoe** [comment]
I realize now this wouldn't work because the MariaDB and RabbitMQ operator need to read the secret. So we would need to still create the secret from this data. So still more to iterate here.

**Comment by skrobul** [comment]
https://github.com/jkroepke/helm-secrets can use sops and it can generate both plain values, Secrets or whatever else. The downside over sealed secrets is that whoever has the key can view the credentials locally. Can be partially mitigated  if we use non-gpg backend

**Comment by cardoe** [comment]
Yeah maybe the docs give a couple of suggested options and that's one?

**Comment by cardoe** [comment]
The goal with this is to make it easier for folks to run the helm commands and to be less opinionated about the secrets storage.

**Comment by cardoe** [comment]
So to provide some more context:
1. remove using Sealed Secrets for OpenStack Helm components, which removes the need for the post-renderer
2. Put secrets into a values file instead of creating them as separate files and then loading them dynamically from the cluster

This should make us less opinionated and easier to use.

**Comment by cardoe** [comment]
Ultimately the helm commands go from:

```bash
helm --namespace openstack template \
     ironic \
    ./openstack-helm/ironic/ \
    -f components/13-ironic/aio-values.yaml \
    --set endpoints.identity.auth.admin.password="$(kubectl --namespace openstack get secret keystone-admin -o jsonpath='{.data.password}' | base64 -d)" \
    --set endpoints.oslo_db.auth.ironic.password="$(kubectl --namespace openstack get secret ironic-db-password -o jsonpath='{.data.password}' | base64 -d)" \
    --set endpoints.oslo_messaging.auth.ironic.password="$(kubectl --namespace openstack get secret ironic-rabbitmq-password -o jsonpath='{.data.password}' | base64 -d)" \
    --set endpoints.identity.auth.ironic.password="$(kubectl --namespace openstack get secret ironic-keystone-password -o jsonpath='{.data.password}' | base64 -d)" \
    | kubectl -n openstack apply -f -
```

to

```bash
helm --namespace openstack template \
     ironic \
    ./openstack-helm/ironic/ \
    -f components/13-ironic/aio-values.yaml \
    -f secret-openstack.yaml \
    | kubectl -n openstack apply -f -
```

I also wrapped up the helm call with `kustomize` since we've done that in every other component.

File: components/10-keystone/README.md [hunk 1]
```diff
@@ -48,16 +25,14 @@ Secrets Reference:
...
-    --post-renderer $(git rev-parse --show-toplevel)/scripts/openstack-helm-sealed-secrets.sh \
-    | kubectl -n openstack apply -f -
+# create secrets yaml file if you're not already storing or providing it differently
+./scripts/gen-os-secrets.sh secret-openstack.yaml
+
+kubectl kustomize \
+  --enable-helm \
+  --load-restrictor LoadRestrictionsNone \
+  components/10-keystone \
+  | kubectl -n openstack apply -f -
```
Comment (skrobul) [inline components/10-keystone/README.md:35]:
```
❯ ./scripts/gen-os-secrets.sh secret-openstack.yaml

This script will attempt to look up the existing values this repo used
or will generate new values. The output below will be related to that.

~/dev/understack ☸ kind-secretstest  pr/19[$?] 
❯ cat secret-openstack.yaml                          
# The purpose of this file is to serve as a template for OpenStack Helm
# based secrets values that are necessary for OpenStack Helm to populate
# it's configmap-etc, which is really a secret with connection strings
---
endpoints:
  # 'identity' endpoints are for keystone access
  identity:
    auth:
      # this is the 'admin' user created in keystone by the initial start
      # and used by the other services to create their service accounts
      # and endpoint in the service catalog.
      admin:
        password: "J3Fugvb2591H18OE89dC0CETsgY1bRUt"
      # this user is the service account that ironic uses
      ironic:
        password: "SPtvZAoVsc5BnJKA3oOtrq1-X5tervUs"
  # 'oslo_db' is for MariaDB
  oslo_db:
    auth:
      # this is what the keystone service uses to connect to MariaDB
      keystone:
        password: "e7uq2fF9IotLgbEjLcW4NJWYptylCBgq"
      # this is what the ironic service uses to connect to MariaDB
      ironic:
        password: "kCgPfbqd_vr1sLUcbw1tTvg9mX5y_at9"
  # 'oslo_messaging' is for RabbitMQ
  oslo_messaging:
    auth:
      # this is what the keystone service uses to connect to RabbitMQ
      keystone:
        password: "BjcIo-f9aF4Yb4E4RACF4beJxbelYQu2"
      # this is what the ironic service uses to connect to RabbitMQ
      ironic:
        password: "HwHoM1zPLW07VYWx4pIjuZa9pkBilmI0"

~/dev/understack ☸ kind-secretstest  pr/19[$?] 
❯ kubectl kustomize \
  --enable-helm \
  --load-restrictor LoadRestrictionsNone \
  components/10-keystone \
  | kubectl -n openstack apply -f -

error: json: unknown field "additionalValuesFiles"
error: no objects passed to apply
```

```
❯ kubectl kustomize \
  --enable-helm \
  --load-restrictor LoadRestrictionsNone \
  components/10-keystone  
error: json: unknown field "additionalValuesFiles"
```

Comment (skrobul) [reply components/10-keystone/README.md:35]: Further debugging:

```
❯ sh -x scripts/gen-os-secrets.sh secret-openstack.yaml
+ '[' 1 -ne 1 ']'
+ set -o pipefail
+ type -p yq
+ type -p kubectl
++ dirname scripts/gen-os-secrets.sh
+ SCRIPTS_DIR=scripts
+ echo 'This script will attempt to look up the existing values this repo used'
This script will attempt to look up the existing values this repo used
+ echo 'or will generate new values. The output below will be related to that.'
or will generate new values. The output below will be related to that.
++ kubectl -n openstack get secret keystone-admin -o 'jsonpath={.data.password}'
++ base64 -d
+ export KEYSTONE_ADMIN_PASSWORD=J3Fugvb2591H18OE89dC0CETsgY1bRUt
+ KEYSTONE_ADMIN_PASSWORD=J3Fugvb2591H18OE89dC0CETsgY1bRUt
++ kubectl -n openstack get secret keystone-db-password -o 'jsonpath={.data.password}'
++ base64 -d
+ export KEYSTONE_DB_PASSWORD=e7uq2fF9IotLgbEjLcW4NJWYptylCBgq
+ KEYSTONE_DB_PASSWORD=e7uq2fF9IotLgbEjLcW4NJWYptylCBgq
++ kubectl -n openstack get secret keystone-rabbitmq-password -o 'jsonpath={.data.password}'
++ base64 -d
+ export KEYSTONE_RABBITMQ_PASSWORD=BjcIo-f9aF4Yb4E4RACF4beJxbelYQu2
+ KEYSTONE_RABBITMQ_PASSWORD=BjcIo-f9aF4Yb4E4RACF4beJxbelYQu2
++ kubectl -n openstack get secret ironic-keystone-password -o 'jsonpath={.data.password}'
++ base64 -d
+ export IRONIC_KEYSTONE_PASSWORD=SPtvZAoVsc5BnJKA3oOtrq1-X5tervUs
+ IRONIC_KEYSTONE_PASSWORD=SPtvZAoVsc5BnJKA3oOtrq1-X5tervUs
++ kubectl -n openstack get secret ironic-db-password -o 'jsonpath={.data.password}'
++ base64 -d
+ export IRONIC_DB_PASSWORD=kCgPfbqd_vr1sLUcbw1tTvg9mX5y_at9
+ IRONIC_DB_PASSWORD=kCgPfbqd_vr1sLUcbw1tTvg9mX5y_at9
++ kubectl -n openstack get secret ironic-rabbitmq-password -o 'jsonpath={.data.password}'
++ base64 -d
+ export IRONIC_RABBITMQ_PASSWORD=HwHoM1zPLW07VYWx4pIjuZa9pkBilmI0
+ IRONIC_RABBITMQ_PASSWORD=HwHoM1zPLW07VYWx4pIjuZa9pkBilmI0
+ yq -M '(.. | select(tag == "!!str")) |= envsubst' scripts/../components/openstack-secrets.tpl.yaml

~/dev/understack ☸ kind-secretstest  pr/19[$!?] 
```
^ produces empty output. interestingly, when ran outside of script, the same `yq` command renders output for me. I'm looking into why

Comment (skrobul) [reply components/10-keystone/README.md:35]: nevermind, wrong kubectl version

File: scripts/gen-os-secrets.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+if ! type -p yq > /dev/null; then
+    echo "You must have yq installed to use this script" >&2
+    exit 1
+fi
+
+if ! type -p kubectl > /dev/null; then
+    echo "You must have kubectl installed to use this script" >&2
+    exit 1
+fi
+
```
Comment (skrobul) [inline scripts/gen-os-secrets.sh:19]:
```suggestion
if ! type -p kubectl > /dev/null; then
    echo "You must have kubectl installed to use this script" >&2
    exit 1
fi

KUSTOMIZE_VERSION=$(kubectl version --client -o yaml | yq .kustomizeVersion)
if ! (echo -e "v5.0.0\n$KUSTOMIZE_VERSION" | sort -V -C); then
  echo "kustomize needs to be at version 5.0.0 or newer (comes with kubectl 1.27+)"
  exit 1
fi

```

`additionalValuesFiles` has been [added](https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv5.0.0) in v5.0.0 so we should probably check for availability of this or newer version

**Review by skrobul — CHANGES_REQUESTED** [review]

**Review by skrobul — APPROVED** [review]
previous review was supposed to be with Approve, not Request changes, sorry

**Review by skrobul — APPROVED** [review]

**Pull request merged**